File: code/sorting/merge_sort/merge_sort.fs
//Part of Cosmos by OpenGenus Foundation
open System
open System.Collections.Generic

let rec merge (left:IComparable list) (right:IComparable list): IComparable list = 
    match (left,right) with
    | (_,[]) -> left
    | ([],_) -> right
    | (l::ls,r::rs) when  l < r -> l::merge ls right
    | (l::ls,r::rs) -> r::merge left rs

let rec merge_sort= 
    function
    | [x]-> [x]
    |  x -> let (left,right) = List.splitAt (x.Length/2) x
            merge (merge_sort left) (merge_sort right)
    

[<EntryPoint>]

let main(argv) =
    printfn "%A" <| merge_sort [10;22;41;2;33;12;52]
    0

File: code/sorting/merge_sort/merge_sort.hs
-- Part of Cosmos by OpenGenus Foundation
split_list::[a]->([a],[a])
split_list x = splitAt (div (length x) $ 2) x

merge_sort::Ord a=>[a]->[a]
merge_sort x
    | length x == 0 = []
    | length x == 1 = x
    | length x > 1 = merge (merge_sort left) (merge_sort right)
        where (left,right)=split_list x
    


merge::Ord a=>[a]->[a]->[a]
merge [] r = r
merge l [] = l
merge l@(l':ls) r@(r':rs) 
    | l' < r' =  l' : merge ls r
    | otherwise = r' : merge rs l

main =  do
    print $ merge_sort [9,6,4,3,2,1,2,5]
